java/src/main/native/src/TableJni.cpp

```diff
@@ -245,6 +248,49 @@ class native_arrow_ipc_writer_handle final {
     }
     initialized = false;
   }
+
+  std::vector<cudf::column_metadata> get_column_metadata(const table_view& tview) {
+    if (!column_names.empty() && columns_meta.empty()) {
+      // Rebuild the structure of column meta according to table schema.
+      // All the tables written by this writer should share the same schema,
+      // so build column metadata only once.
+      columns_meta.reserve(tview.num_columns());
+      size_t idx = 0;
+      for (auto itr = tview.begin(); itr < tview.end(); ++itr) {
+        columns_meta.push_back(std::move(build_one_column_meta(*itr, idx)));
+        idx ++;
+      }
+    }
```

Comment:
There should be a check here for `idx == column_names.size()` in case the user passed more names than we ended up using which would be surprising to the caller.

Reply:
good suggestion, updated